### PR TITLE
Add a client for Virtual Network management

### PR DESCRIPTION
**Pull request by jen20**
This pull requests support for managing virtual networks in Azure.

It adds structs to parse and generate the (seemingly unavailable) XML schema for virtual network specification, and a client with Get and Set operations for the currently active subscription.

The pattern for using the client is the same as the powershell version - get the running config, merge in any desired changes and set the configuration to the new state. There is no documentation on what the concurrency constraints are with the underlying Azure REST API, I suspect it is not safe to call concurrently though can't verify this.

It was also necessary to modify the HTTP client to accept different content types, since the underlying REST API requires the PUT operation to be made with a content type of `text/plain`, despite it actually being XML.

### CODE REVIEW

File: clients/vnetClient/vnetClient.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+//Note that the underlying Azure API means that network related operations
+//are not safe for running concurrently.
+func SetVirtualNetworkConfiguration(networkConfiguration *NetworkConfiguration) error {
+	networkConfiguration.setXmlNamespaces()
+	networkConfigurationBytes, err := xml.Marshal(networkConfiguration)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Content", string(networkConfigurationBytes))
```
Comment (hopetobelievein) [inline clients/vnetClient/vnetClient.go]: Do you need this?

Comment (jen20) [inline clients/vnetClient/vnetClient.go]: Nope, that was an oversight from when I was trying to debug the schema. I'll remove.

**Comment by ahmetalpbalkan** [comment]
Yeah API doesn't support concurrency. :) It's easy to run into race conditions in VNet REST APIs. Feel free to upvote my feedback: http://feedback.azure.com/forums/217313-networking-dns-traffic-manager-vpn-vnet/suggestions/5954992-make-virtual-network-operations-atomic-on-portal-a

File: clients/vnetClient/vnetClient.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+const (
+	azureNetworkConfigurationURL = "services/networking/media"
+)
+
+//GetVirtualNetworkConfiguration retreives the current virtual network
+//configuration for the currently active subscription. Note that the
+//underlying Azure API means that network related operations are not safe
+//for running concurrently.
+func GetVirtualNetworkConfiguration() (*NetworkConfiguration, error) {
```
Comment (ahmetalpbalkan) [inline clients/vnetClient/vnetClient.go]: shouldn't it be just NetworkConfiguration, without pointer?

File: clients/vnetClient/vnetClient.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+	return networkConfiguration, nil
+}
+
+//SetVirtualNetworkConfiguration configures the virtual networks for the
+//currently active subscription according to the NetworkConfiguration given.
+//Note that the underlying Azure API means that network related operations
+//are not safe for running concurrently.
+func SetVirtualNetworkConfiguration(networkConfiguration *NetworkConfiguration) error {
+	networkConfiguration.setXmlNamespaces()
```
Comment (ahmetalpbalkan) [inline clients/vnetClient/vnetClient.go]: we can probably guarantee they are already set. see my next comment above.

File: clients/vnetClient/vnetClient.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+//configuration for the currently active subscription. Note that the
+//underlying Azure API means that network related operations are not safe
+//for running concurrently.
+func GetVirtualNetworkConfiguration() (*NetworkConfiguration, error) {
+	response, err := azure.SendAzureGetRequest(azureNetworkConfigurationURL)
+	if err != nil {
+		return nil, err
+	}
+
+	networkConfiguration := &NetworkConfiguration{}
```
Comment (ahmetalpbalkan) [inline clients/vnetClient/vnetClient.go]: use NewNetworkConfiguration here first, then unmarshal. that way you'll return an object with XML namespaces set. although if constructing this object from scratch is a valid use case, we'll still need setXmlNamespaces.

Comment (jen20) [inline clients/vnetClient/vnetClient.go]: Could do that - whether constructing from scratch is valid really depends on use case though. Leaving the call in there seems safer though.

File: clients/vnetClient/vnetClient.go (hunk 2 above)
Comment (jen20) [inline clients/vnetClient/vnetClient.go]: Can see arguments for either here. Will look through what other clients do and potentially change to returning the value instead.

**Comment by jen20** [comment]
@ahmetalpbalkan re: concurrency a reasonable way to do this would be to support `If-Match` on the tag of the configuration. I suspect that will not be supported by Azure any time soon though!

File: clients/vnetClient/vnetClient.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	}
+
+	return networkConfiguration, nil
+}
+
+//SetVirtualNetworkConfiguration configures the virtual networks for the
+//currently active subscription according to the NetworkConfiguration given.
+//Note that the underlying Azure API means that network related operations
+//are not safe for running concurrently.
+func SetVirtualNetworkConfiguration(networkConfiguration *NetworkConfiguration) error {
```
Comment (hopetobelievein) [inline clients/vnetClient/vnetClient.go]: Check for null param here?

Comment (ahmetalpbalkan) [inline clients/vnetClient/vnetClient.go]: this should not be a pointer anyway.

**Comment by hopetobelievein** [comment]
Can you fix Travis build please?

**Comment by ahmetalpbalkan** [comment]
rebasing the commit on top of master should be fixing the problem.

**Comment by jen20** [comment]
Yes, rebasing and changing the use of unexported symbols fixes this.

**Comment by jen20** [comment]
By the way, what's the thought in this project of squashing commits prior to PR? Prefer the clean history or the complete change sets?

**Comment by ahmetalpbalkan** [comment]
For changes essentially changing one major thing, one squashed commit preferable. Thanks.

**Comment by jen20** [comment]
Do you want me to squash this one?

**Comment by ahmetalpbalkan** [comment]
Yes please.

**Comment by ahmetalpbalkan** [comment]
LGTM

**Comment by jen20** [comment]
:+1: in use it's actually a bit derpy because of the semantics around modifying the "document" that gets returned when it's not a pointer. May revisit that in future.

**Pull request merged**